src/coreclr/jit/helperexpansion.cpp

```diff
@@ -457,7 +459,196 @@ PhaseStatus Compiler::fgExpandThreadLocalAccess()
+    DebugInfo   debugInfo    = stmt->GetDebugInfo();
+    block                    = fgSplitBlockBeforeTree(block, stmt, call, &newFirstStmt, &callUse);
+    *pBlock                  = block;
+    var_types callType       = call->TypeGet();
+    assert(prevBb != nullptr && block != nullptr);
+
+    unsigned finalLclNum         = lvaGrabTemp(true DEBUGARG("Final offset"));
+    lvaTable[finalLclNum].lvType = TYP_I_IMPL;
+    GenTree* finalLcl            = gtNewLclVarNode(finalLclNum);
+
+    // Block ops inserted by the split need to be morphed here since we are after morph.
+    // We cannot morph stmt yet as we may modify it further below, and the morphing
+    // could invalidate callUse.
+    while ((newFirstStmt != nullptr) && (newFirstStmt != stmt))
+    {
+        fgMorphStmtBlockOps(block, newFirstStmt);
+        newFirstStmt = newFirstStmt->GetNextStmt();
+    }
+
+    if (TargetOS::IsWindows)
+    {
+#ifdef TARGET_64BIT
+        // prevBb (BBJ_NONE):                                               [weight: 1.0]
+        //      ...
+        //
+        // tlsRootNullCondBB (BBJ_COND):                                    [weight: 1.0]
+        //      fastPathValue = [tlsRootAddress]
+        //      if (fastPathValue != nullptr)
+        //          goto fastPathBb;
+        //
+        // fallbackBb (BBJ_ALWAYS):                                         [weight: 0]
+        //      tlsRoot = HelperCall();
+        //      goto block;
+        //
+        // fastPathBb(BBJ_ALWAYS):                                          [weight: 1.0]
+        //      tlsRoot = fastPathValue;
+        //
+        // block (...):                                                     [weight: 1.0]
+        //      use(tlsRoot);
+        // ...
+
+        // Mark this ICON as a TLS_HDL, codegen will use FS:[cns] or GS:[cns]
+        GenTree* tlsValue = gtNewIconHandleNode(threadStaticInfo.offsetOfThreadLocalStoragePointer, GTF_ICON_TLS_HDL);
+        tlsValue          = gtNewIndir(TYP_I_IMPL, tlsValue, GTF_IND_NONFAULTING | GTF_IND_INVARIANT);
+
+        CORINFO_CONST_LOOKUP tlsIndexObject = threadStaticInfo.tlsIndexObject;
+
+        GenTree* dllRef = gtNewIconHandleNode((size_t)tlsIndexObject.handle, GTF_ICON_OBJ_HDL);
+        dllRef          = gtNewIndir(TYP_INT, dllRef, GTF_IND_NONFAULTING | GTF_IND_INVARIANT);
+        dllRef          = gtNewOperNode(GT_MUL, TYP_I_IMPL, dllRef, gtNewIconNode(TARGET_POINTER_SIZE, TYP_INT));
+
+        // Add the dllRef to produce thread local storage reference for coreclr
+        tlsValue = gtNewOperNode(GT_ADD, TYP_I_IMPL, tlsValue, dllRef);
+
+        // Base of coreclr's thread local storage
+        tlsValue = gtNewIndir(TYP_I_IMPL, tlsValue, GTF_IND_NONFAULTING | GTF_IND_INVARIANT);
+
+        CORINFO_CONST_LOOKUP tlsRootObject = threadStaticInfo.tlsRootObject;
+
+        GenTree* tlsRootOffset = gtNewIconNode((size_t)tlsRootObject.handle, TYP_INT);
+        tlsRootOffset->gtFlags |= GTF_ICON_SECREL_OFFSET;
```

Comment:
Shouldn't this use `gtNewIconHandleNode`?

Reply:
no, this resolves to an immediate offset itself that gets relocated.

JITDisasm:
```asm
       mov      eax, dword ptr [(reloc 0x40000000004229f0)]
       mov      edx, (reloc 0x40000000004229f8)  ;<-- offset
       add      rdx, qword ptr [rcx+8*rax]
       mov      rbx, qword ptr [rdx]
       test     rbx, rbx
       je       SHORT G_M35921_IG06
```

Dumpbin:

```asm
  00000000000B15DC: 8B 05 00 00 00 00  mov         eax,dword ptr [_tls_index]
  00000000000B15E2: BA 00 00 00 00     mov         edx,offset tls_InlinedThreadStatics ;<-- offset
  00000000000B15E7: 48 03 14 C1        add         rdx,qword ptr [rcx+rax*8]
  00000000000B15EB: 48 8B 1A           mov         rbx,qword ptr [rdx]
  00000000000B15EE: 48 85 DB           test        rbx,rbx
  00000000000B15F1: 74 34              je          00000000000B1627
```

Assembly:

```asm
00007FF72C8405FC  mov         eax,dword ptr [_tls_index (07FF72CA6ADCCh)]  
00007FF72C840602  mov         edx,10h   ;<-- offset
00007FF72C840607  add         rdx,qword ptr [rcx+rax*8]  
00007FF72C84060B  mov         rbx,qword ptr [rdx]  
00007FF72C84060E  test        rbx,rbx  
00007FF72C840611  je          repro_Program__Test2+67h (07FF72C840647h) 
```